### PR TITLE
[ALLUXIO-1753] Extend the copyFromLocal command and add the support of folder copy

### PR DESCRIPTION
**Pull request by ConeyLiu**
https://alluxio.atlassian.net/browse/ALLUXIO-1753

### CODE REVIEW

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 1]
```diff
@@ -170,7 +170,15 @@ private void copyPath(File src, AlluxioURI dstPath) throws IOException {
           closer.close();
         }
       } else {
-        mFileSystem.createDirectory(dstPath);
+        // Add a judgement, because the dstPath maybe already exists. For example:
+        // 'copyFromLocal pathToSrcDir /'.
+        try {
+          mFileSystem.createDirectory(dstPath);
+        } catch (FileAlreadyExistsException e) {
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: Why is this needed? I think it makes more sense to notify the user that the file already exists.

Comment (ConeyLiu) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: @yupeng9  I think it is common to copy a directory from local to existing directory. And even you run `bin/alluxio copyFromLocal srcDir /`, it only return you `/` . not 'the file already exits' .

Comment (ConeyLiu) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: @yupeng9  Sorry, I see what you mean. I'll fix it.

Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: By design, `copyFromLocal` works differently like unix `cp`, when you want to copy srcDir under dstDir, you need to say `copyFromLocal srcDir dstDir/srcDir`.

**Comment by ConeyLiu** [comment]
@uronce-cc @yupeng9 Thanks for your comments, and I have modified. Please verify it. Thanks again.

**Comment by ConeyLiu** [comment]
```java
 /**
   * Copies a directory from local to dstPath in the Alluxio. The destination directory structure
   * maintained as local directory. This method is used when input path is a directory.
   * @param srcDir The directory in the local filesystem.
   * @param dstPath The {@link AlluxioURI} of the destination.
   * @throws IOException If a non-Alluxio related exception occurs.
   */
```
Who can tell me what is the wrong?  There still a  error : `Period is not needed at the end of the @param tag.` .

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 2]
```diff
@@ -224,8 +224,8 @@ public InodeDirectory getRoot() {
   public CreatePathResult createPath(AlluxioURI path, CreatePathOptions options)
       throws FileAlreadyExistsException, BlockInfoException, InvalidPathException, IOException {
     if (path.isRoot()) {
-      LOG.info("FileAlreadyExistsException: {}", path);
-      throw new FileAlreadyExistsException(path.toString());
+      LOG.info(ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path));
+      throw new FileAlreadyExistsException(ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path));
```
Comment (yupeng9) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: thanks for fixing this.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 3]
```diff
@@ -80,15 +79,75 @@ public void run(CommandLine cl) throws IOException {
...
+      }
+    }
+    if (errorMessages.size() != 0) {
+      throw new IOException(Joiner.on('\n').join(errorMessages));
+    }
+  }
+
+  /**
+   * Create a directory in the Alluxio filesystem space. It will not throw any exception if the
+   * destination directory is already exists.
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: grammar: directory already exists

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 4]
```diff
@@ -80,15 +79,75 @@ public void run(CommandLine cl) throws IOException {
...
+            copyFromLocalDirs(srcFile, new AlluxioURI(dstPath, new AlluxioURI(srcFile.getName())));
+          } else {
+            copyPath(srcFile, new AlluxioURI(dstPath, new AlluxioURI(srcFile.getName())));
+          }
+        } catch (IOException e) {
+          errorMessages.add(e.getMessage());
+        }
+      }
+      if (errorMessages.size() != 0) {
+        if (errorMessages.size() == fileList.length) {
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: I feel this check is not sufficient to tell no file is created: it's possible that some files are created in the subdirectory from the call `copyFromLocalDirs` and then failed there, right?

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 5]
```diff
@@ -80,15 +79,75 @@ public void run(CommandLine cl) throws IOException {
...
+          // If no files were created, then delete the directory
+          if (mFileSystem.exists(dstPath)) {
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: I think this cleanup is not necessary if some failures happens in the subtree copy.

File: tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java [hunk 6]
```diff
@@ -93,6 +93,25 @@ public void copyFromLocalFileToDstPathTest() throws IOException, AlluxioExceptio
   }
 
   @Test
+  public void copyFromLocalDirToExistingDirTest() throws IOException, AlluxioException {
+    File srcOuterDir = new File(mLocalAlluxioCluster.getAlluxioHome() + "/outerDir");
+    File srcInnerDir = new File(mLocalAlluxioCluster.getAlluxioHome() + "/outerDir/innerDir");
+    srcOuterDir.mkdir();
+    srcInnerDir.mkdir();
+    generateFileContent("/outerDir/srcFile1", BufferUtils.getIncreasingByteArray(10));
+    generateFileContent("/outerDir/innerDir/srcFile2", BufferUtils.getIncreasingByteArray(10));
```
Comment (yupeng9) [inline tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java]: Could you also create an empty subfolder, and check if it's copied as well?

**Comment by yupeng9** [comment]
Thanks for extending the copyFromLocal command and adding the support of folder copy. I left some comments.

btw, the title of this PR is no longer correct, please update it accordingly. And also please update the JIRA ticket's description.

File: tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java (hunk 6 above)
Comment (ConeyLiu) [inline tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java]: Now it‘ll copy an empty subfolder, should I change it?

**Comment by ConeyLiu** [comment]
@yupeng9 Thanks for you comments. I have fixed the problem which you commented. Please verify it.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 7]
```diff
@@ -80,15 +79,80 @@ public void run(CommandLine cl) throws IOException {
...
+   * @param dstPath The {@link AlluxioURI} of the destination
+   * @throws IOException if a non-Alluxio related exception occurs
+   *
+   */
+  private void copyFromLocalDirs(File srcDir, AlluxioURI dstPath) throws IOException {
+    try {
+      createDscDir(dstPath);
+      List<String> errorMessages = Lists.newArrayList();
+      File[] fileList = srcDir.listFiles();
+      int misFiles = 0;
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: The convention of having prefix `m` is for class member variables.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 8]
```diff
@@ -80,15 +79,80 @@ public void run(CommandLine cl) throws IOException {
...
+          errorMessages.add(e.getMessage());
+          if (!mFileSystem.exists(newURI)) {
+            misFiles++;
+          }
+        }
+      }
+      if (errorMessages.size() != 0) {
+        if (misFiles == fileList.length) {
+          // If no files were created, then delete the directory
+          if (mFileSystem.exists(dstPath)) {
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: But what if the dstPath already exists before the copy, and has files under it?

**Comment by yupeng9** [comment]
Thanks for addressing the comments, and adding the tests. I added some more comments.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 7 above)
Comment (ConeyLiu) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: The `misFiles` is just a local variable, should it change?

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 8 above)
Comment (ConeyLiu) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: Thanks for the comment. I have fixed it, and update the unit test.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 9]
```diff
@@ -80,15 +79,81 @@ public void run(CommandLine cl) throws IOException {
...
   /**
+   * Copies a directory from local to Alluxio filesystem. The destination directory structure
+   * maintained as local directory. This method is used when input path is a directory.
+   *
+   * @param srcDir The source directory in the local filesystem
+   * @param dstPath The {@link AlluxioURI} of the destination
+   * @throws IOException if a non-Alluxio related exception occurs
+   *
+   */
+  private void copyFromLocalDirs(File srcDir, AlluxioURI dstPath) throws IOException {
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: copyFromLocalDir

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 10]
```diff
@@ -80,15 +79,81 @@ public void run(CommandLine cl) throws IOException {
...
+    try {
+      boolean existing = mFileSystem.exists(dstPath);
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: make this more descriptive, e.g. `dstExistedBefore`.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 11]
```diff
@@ -80,15 +79,81 @@ public void run(CommandLine cl) throws IOException {
...
+          }
+        } catch (IOException e) {
+          errorMessages.add(e.getMessage());
+          if (!mFileSystem.exists(newURI)) {
+            misFiles++;
+          }
+        }
+      }
+      if (errorMessages.size() != 0) {
+        if (misFiles == fileList.length) {
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: we should delete the directory if there are any errors, no need for this check.

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 12]
```diff
@@ -80,15 +79,81 @@ public void run(CommandLine cl) throws IOException {
...
+    }
+  }
+
+  /**
+   * Create a directory in the Alluxio filesystem space. It will not throw any exception if the
+   * destination directory already exists.
+   * @param dstPath The {@link AlluxioURI} of the destination directory which will be created
+   * @throws IOException if a non-Alluxio related exception occurs
+   */
+  private void createDscDir(AlluxioURI dstPath) throws IOException {
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: createDscDir -> createDstDir

File: tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java [hunk 13]
```diff
@@ -93,6 +145,30 @@ public void copyFromLocalFileToDstPathTest() throws IOException, AlluxioExceptio
...
+    File srcOuterDir = new File(mLocalAlluxioCluster.getAlluxioHome() + "/outerDir");
+    File srcInnerDir = new File(mLocalAlluxioCluster.getAlluxioHome() + "/outerDir/innerDir");
+    File emptyDir = new File(mLocalAlluxioCluster.getAlluxioHome() + "/outerDir/emptyDir");
+    srcOuterDir.mkdir();
+    srcInnerDir.mkdir();
+    emptyDir.mkdir();
+    generateFileContent("/outerDir/srcFile1", BufferUtils.getIncreasingByteArray(10));
+    generateFileContent("/outerDir/innerDir/srcFile2", BufferUtils.getIncreasingByteArray(10));
+    mFileSystem.createDirectory(new AlluxioURI("/dstDir"));
+    int ret = mFsShell.run("copyFromLocal", srcOuterDir.getPath() + "/", "/dstDir");
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/shell/command/CopyFromLocalCommandTest.java]: is the added "/" necessary? Ideally it shouldn't be.

**Comment by aaudiber** [comment]
Thanks for the PR, this is useful functionality to have. I've left a few comments on the code.

Please add comments for each unit test to say what case(s) it is testing. Some cases that should be tested are
1. copying a directory to a destination directory which exists and doesn't contain the copied directory.
```
-src_dir
-dst_dir
--not_src_dir
cp src_dir dst_dir
```
should succeed
2. copying a directory to a destination directory which exists and does contain the copied directory.
```
-src_dir
-dst_dir
--src_dir
cp src_dir dst_dir
```
should fail
3. copying a directory to a destination directory which does not exist.
```
-src_dir
-not_dst_dir
cp src_dir dst_dir
```
should succeed
4. copying a directory to a destination file.
```
-src_dir
-dst_file
cp src_dir dst_file
```
should fail

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 11 above)
Comment (ConeyLiu) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: @aaudiber If some files copied successful, but only one copied failed. Just as this situation, I think the directory shouldn't be deleted.

**Comment by ConeyLiu** [comment]
@aaudiber Thanks for your comments, I have fixed the problem which you comments. Please check it.

**Comment by yupeng9** [comment]
LGTM. Thanks for the improvement!

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 11 above)
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: that makes sense, sgtm

**Comment by aaudiber** [comment]
+1 LGTM
